requests/models.py

```
@@ -155,8 +155,12 @@ def _encode_files(files, data):
 
             if isinstance(fp, (str, bytes, bytearray)):
                 fdata = fp
-            else:
+            elif hasattr(fp, 'read'):
                 fdata = fp.read()
+            elif fp is None:
+                continue
+            else:
```

In this case, are we expecting the objects `__str__` representation to be sufficient? We currently blow up on things like `datetime` or `dict`, but this may have unexpected results.

If we're OK with the above, would it make more sense to reorg the conditionals to check for `read`, `None` and then everything else would be `fdata = fp`?

e.g.
```python
if hasattr(fp, 'read'):
    fdata = fp.read()
elif fp is None:
    continue
else:
    fdata = fp
```

---

So the code that will eventually handle this is [here](https://github.com/urllib3/urllib3/blob/c8c98c473ab30f427fe05057f2ceded10ec93153/urllib3/filepost.py#L78..L90). 

Honestly, at this point, we *could* simplify the conditional but I'd rather leave it and let it blow up for folks. In 3.0, I'd love to have this else branch raise an exception, though.

---

Alright, I'm on board with that. @mayani would you mind removing this `else` conditional to keep our existing functionality. We'll look at adding a more explicit exception in 3.0. Thanks!

---

@nateprewitt 

```python
else:
    fdata = fp

rf = RequestField(name=k, data=fdata, filename=fn, headers=fh)
```

Getting rid of else part will result in fdata being an unbound variable. Should we set it to fdata=None in else and let it blow up or leave things as is?

---

Urgh, yep. Ok so @sigmavirus24 we may just want to put the exception in now. Non-file objects were already exploding on the read call so they never worked.

We *could* make sure the exception inherits from AttributeError if we want to ensure we’re not breaking anyone currently relying on this defect for flow control.

---

@nateprewitt I'd rather never inherit from `AttributeError` if possible. That is just a world of pain. We've done this in the past, but that means if people are putting too much code in their `try/except` block, we may cause them to enter an exception handling block.

In short, I think the code as is may be safe right now. If we want to be extra magical, we could do `fdata = str(fp)` but I really don't want to add that behaviour at all. Let's just leave this as is, I guess and plan to add the exception in 3.0

---

Sounds good! I think I misread your first response, so we should be set then.